### DISCUSSION TITLE
Move files by keyword

### DISCUSSION THREAD

**Discussion by matiaseche** [rubyatscale/packs#75] [General]
Hello :wave: 

I've been working a lot with modularization lately in a big monolith, using all the tools available at `rubyatscale`. Thanks for developing and maintaining such useful tools!

This particular gem, `use_packs` I have not yet used, because it feels repetitive.
For the codebase I am working on, I wrote a bash script to create packs which follows these pattern:
```bash
bin/create_pack -p <pack_name> -k <keyword> [-f <folder>]
```
What this does is:

1. Create the pack in the dir `pack_name`, if it already exists then it goes to the next step.
2. Move **every file** that contains the keyword given on the `-k` flag (except for a list of directories like "db/", "public/", "bin", "packs", "fixtures" and others)
    * If the optional flag `-f` is given, only files under `folder` **and** containing the keyword in their path will be moved.

So for example:
```bash
bin/create_pack -p packs/first_pack -k namespace/
```

Will move all files that have `namespace/` in their path to the created pack, which eases the process of moving files and is less repetitive.

With these approach all files under `app/**/namespace/*`, `config/locales/**/namespace/*` and `test/**/namespace/*` will be moved.
This does not guarantee that all the files moved are the ones that should be on the pack, but it sure moves most. Then after reviewing the changes the developer can note if a file is not supposed to be there and revert that change.

Using `use_packs` it would have been the same command, but repeat for models, controlles, views, helpers, jobs, mailers, and whatever other folders are in the app, and then the same for tests.

Does this make sense? Would it be useful to add a functionality to support this to the gem? What approaches have you used to ease this process?

**Comment by alexevanczuk** [comment]
Hey @matiaseche thanks for the kind words and providing feedback. We actually started with the same CLI tools too (`bin/create_pack`, for example).

The gem currently doesn't have much tooling to find files to move based on any heuristics. The only thing it can do at the moment is accept files and folders. One small addition that maybe could help is having it accept globs, but I chose not to add that since you can always pass in an expanded glob into the public API:
```ruby
UsePacks.move_to_pack!(
    pack_name: 'packs/my_pack',
    paths_relative_to_root: paths_relative_to_root,
)
```

So in this case, you could use the public API and pass in an identical glob, i.e.
```ruby
paths_relative_to_root = Dir.glob('**/namespace/**')
```

If you wanted to pass in files that contain a certain string, you could do:
```ruby
paths_relative_to_root = Pathname.glob('**/**').select{|p| p.read.include?('my_keyword')}
```

I know it's not quite as convenient as being able to do it on the command line, but I imagine that big messy moves involving heuristic based matching probably happen a bit more rarely than surgical moves of single files or folders. 

Let me know what ya think!

**Reply by matiaseche** [reply]
Thanks for the reply!

I understand why it is not supported, and is good to know that an expanded glob can be passed into the public API.

Still as of now, for the project I am working on, the script serves its purpose, and is not as messy since most teams have their own namespaces :+1:
